### PR TITLE
Added option to enable/disable envoy route

### PR DESCRIPTION
**Pull request by hbelmiro**
## The issue resolved by this Pull Request:
Resolves https://issues.redhat.com/browse/RHOAIENG-6463

## Description of your changes:

This PR:

- Fixes tests in mlmd_test.go
  - Changes existing tests to cover v1 and v2 
- Adds option to enable/disable envoy route

## Testing instructions

There are 6 scenarios to cover (This PR adds unit tests for all of them):

| Scenario | DSP Version | Configuration | Envoy route deployed |
| ------------ |------------ | --- | --- |
| 1 | V2 | Envoy route enabled | Yes |
| 2 | V2 | Envoy route disabled | No |
| 3 | V2 | Default MLMD| Yes |
| 4 | V1 | Envoy route enabled | Yes |
| 5 | V1 | Envoy route disabled | No |
| 6 | V1 | Default MLMD | No |

### Scenario 1 - V2 with envoy route enabled

- Deploy the following DSPA:
```yaml
apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
kind: DataSciencePipelinesApplication
metadata:
  name: sample
spec:
  dspVersion: v2
  mlmd:
    deploy: true
    envoy:
      deployRoute: true
  objectStorage:
    minio:
      deploy: true
      image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
  mlpipelineUI:
    image: quay.io/opendatahub/ds-pipelines-frontend:latest
```

- `oc get routes` must show an envoy route.

### Scenario 2 - V2 with envoy route disabled

- Deploy the following DSPA:
```yaml
apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
kind: DataSciencePipelinesApplication
metadata:
  name: sample
spec:
  dspVersion: v2
  mlmd:
    deploy: true
    envoy:
      deployRoute: false
  objectStorage:
    minio:
      deploy: true
      image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
  mlpipelineUI:
    image: quay.io/opendatahub/ds-pipelines-frontend:latest
```

- `oc get routes` mustn't show an envoy route.

### Scenario 3 - V2 with default MLMD

- Deploy the following DSPA:
```yaml
apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
kind: DataSciencePipelinesApplication
metadata:
  name: sample
spec:
  dspVersion: v2
  objectStorage:
    minio:
      deploy: true
      image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
  mlpipelineUI:
    image: quay.io/opendatahub/ds-pipelines-frontend:latest
```

- `oc get routes` must show an envoy route.

### Scenario 4 - V1 with envoy route enabled

- Deploy the following DSPA:
```yaml
apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
kind: DataSciencePipelinesApplication
metadata:
  name: sample
spec:
  mlmd:
    deploy: true
    envoy:
      deployRoute: true
  objectStorage:
    minio:
      deploy: true
      image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
  mlpipelineUI:
    image: quay.io/opendatahub/ds-pipelines-frontend:latest
```

- `oc get routes` must show an envoy route.

### Scenario 5 - V1 with envoy route disabled

- Deploy the following DSPA:
```yaml
apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
kind: DataSciencePipelinesApplication
metadata:
  name: sample
spec:
  mlmd:
    deploy: true
    envoy:
      deployRoute: false
  objectStorage:
    minio:
      deploy: true
      image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
  mlpipelineUI:
    image: quay.io/opendatahub/ds-pipelines-frontend:latest
```

- `oc get routes` mustn't show an envoy route.

### Scenario 6 - V1 with default MLMD

- Deploy the following DSPA:
```yaml
apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
kind: DataSciencePipelinesApplication
metadata:
  name: sample
spec:
  objectStorage:
    minio:
      deploy: true
      image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
  mlpipelineUI:
    image: quay.io/opendatahub/ds-pipelines-frontend:latest
```

- `oc get routes` mustn't show an envoy route.

## Checklist
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by dsp-developers** [comment]
A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-639`
An OCP cluster where you are logged in as cluster admin is required.

To use this image run the following:

```bash
cd $(mktemp -d)
git clone git@github.com:opendatahub-io/data-science-pipelines-operator.git
cd data-science-pipelines-operator/
git fetch origin pull/639/head
git checkout -b pullrequest f9e5070116e42ec0c460a259cd6d87ffdffb7cb5
oc new-project opendatahub
make deploy IMG="quay.io/opendatahub/data-science-pipelines-operator:pr-639"
```

More instructions [here](https://github.com/opendatahub-io/data-science-pipelines-operator#deploy-dsp-instance) on how to deploy and test a Data Science Pipelines Application.

**Review by DharmitD — COMMENTED** [review]
Tested out the first three scenarios:

Output for Scenario 1 - V2 with envoy route enabled:
```
oc get routes
NAME                                HOST/PORT                                                                                               PATH   SERVICES                            PORT           TERMINATION          WILDCARD
ds-pipeline-metadata-envoy-sample   ds-pipeline-metadata-envoy-sample-helber-envoy-true.apps.rosa.ddalvi-latest.al5q.p3.openshiftapps.com          ds-pipeline-metadata-envoy-sample   oauth2-proxy   reencrypt/Redirect   None
ds-pipeline-sample                  ds-pipeline-sample-helber-envoy-true.apps.rosa.ddalvi-latest.al5q.p3.openshiftapps.com                         ds-pipeline-sample                  oauth          reencrypt/Redirect   None
ds-pipeline-ui-sample               ds-pipeline-ui-sample-helber-envoy-true.apps.rosa.ddalvi-latest.al5q.p3.openshiftapps.com                      ds-pipeline-ui-sample               8443           reencrypt/Redirect   None
```
envoy route enabled as expected.

Output for Scenario 2 - V2 with envoy route disabled:
```
 oc get routes
NAME                    HOST/PORT                                                                             PATH   SERVICES                PORT    TERMINATION          WILDCARD
ds-pipeline-sample      ds-pipeline-sample-helber-dspa.apps.rosa.ddalvi-latest.al5q.p3.openshiftapps.com             ds-pipeline-sample      oauth   reencrypt/Redirect   None
ds-pipeline-ui-sample   ds-pipeline-ui-sample-helber-dspa.apps.rosa.ddalvi-latest.al5q.p3.openshiftapps.com          ds-pipeline-ui-sample   8443    reencrypt/Redirect   None
```
Envoy route doesn't come up as expected.
Output for Scenario 3 - V2 with default MLMD:

```
oc get routes
NAME                                HOST/PORT                                                                                                 PATH   SERVICES                            PORT           TERMINATION          WILDCARD
ds-pipeline-metadata-envoy-sample   ds-pipeline-metadata-envoy-sample-helber-mlmd-default.apps.rosa.ddalvi-latest.al5q.p3.openshiftapps.com          ds-pipeline-metadata-envoy-sample   oauth2-proxy   reencrypt/Redirect   None
ds-pipeline-sample                  ds-pipeline-sample-helber-mlmd-default.apps.rosa.ddalvi-latest.al5q.p3.openshiftapps.com                         ds-pipeline-sample                  oauth          reencrypt/Redirect   None
ds-pipeline-ui-sample               ds-pipeline-ui-sample-helber-mlmd-default.apps.rosa.ddalvi-latest.al5q.p3.openshiftapps.com                      ds-pipeline-ui-sample               8443           reencrypt/Redirect   None
```
envoy route comes up as expected.

Since all scenarios work as expected with v2, and the unit tests are functioning, I'm presuming the scenarios work with v1 as well.

/lgtm
Great work @hbelmiro !

**Comment by hbelmiro** [comment]
@diegolovison can you please take a look?

**Comment by diegolovison** [comment]
Why do we need enable/disable?

**Comment by DharmitD** [comment]
> Why do we need enable/disable?

@diegolovison we might not always need to deploy an envoy route, this option provides flexibility from an end user perspective.
For instance, sometimes, deploying the envoy route causes issues with character limit, since the route name is presently too long. This is being tracked in a different issue though, but for now, it would be nice to have a workaround where we get the option to not spin up the envoy route iself.

**Comment by HumairAK** [comment]
still need code review

File: controllers/mlmd.go [hunk 1]
```diff
@@ -33,16 +33,37 @@ func (r *DSPAReconciler) ReconcileMLMD(dsp *dspav1alpha1.DataSciencePipelinesApp
...
 	if params.UsingV1Pipelines(dsp) {
-		err = r.ApplyDir(dsp, params, mlmdTemplatesDir+"/v1")
+		if dsp.Spec.MLMD != nil {
+			if dsp.Spec.MLMD.Envoy == nil || dsp.Spec.MLMD.Envoy.DeployRoute {
+				err := r.ApplyDir(dsp, params, mlmdTemplatesDir)
+				if err != nil {
+					return err
+				}
+			} else {
+				err := r.ApplyDirExcept(dsp, params, mlmdTemplatesDir, []string{mlmdTemplatesDir + "/metadata-envoy.route.yaml.tmpl"})
```
Comment (gmfrasca) [inline controllers/mlmd.go:44]: the logic here checks out, but for consistency we handled this in APIServer by simply putting the route YAML in its own directory and applying it if the `APIServer.EnableRoute` bool was true.  

I worry that this is more complicated than it needs to be, as there are now 'magic templates' that are conditionally applied vs other templates in the same directory, with no way to discern this w/o knowing the code's interal control flow.

APIServer control flow for comparison: https://github.com/opendatahub-io/data-science-pipelines-operator/blob/main/controllers/apiserver.go#L56-L60

File: controllers/mlmd_test.go [hunk 2]
```diff
@@ -18,14 +18,15 @@ limitations under the License.
...
+	v1 "github.com/openshift/api/route/v1"
 	"testing"
 
 	dspav1alpha1 "github.com/opendatahub-io/data-science-pipelines-operator/api/v1alpha1"
 	"github.com/stretchr/testify/assert"
 	appsv1 "k8s.io/api/apps/v1"
 )
 
-func TestDeployMLMD(t *testing.T) {
+func TestDeployMLMDV1(t *testing.T) {
```
Comment (gmfrasca) [inline controllers/mlmd_test.go:29]: we have an upcoming item to make v2 default, so perhaps now that we have versioned tests we should also update the DSPA spec used here to explicitly specify DSPVersion: "v1"

File: controllers/mlmd.go (hunk 1 above)
Comment (hbelmiro) [reply controllers/mlmd.go]: Yeah. Way better.
Done.

**Comment by dsp-developers** [comment]
Change to PR detected. A new PR build was completed.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-639`

File: controllers/mlmd_test.go (hunk 2 above)
Comment (hbelmiro) [reply controllers/mlmd_test.go:29]: Done.

**Comment by DharmitD** [comment]
@hbelmiro tried deploying these changes, 

For case 1 for instance, I just deployed case1, v2 with envoy enabled, but I don't see mlmd route coming up:

```
cat dsp-envoy-enabled.yaml 
apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
kind: DataSciencePipelinesApplication
metadata:
  name: sample
spec:
  dspVersion: v2
  mlmd:
    deploy: true
    envoy:
      deployRoute: true
  objectStorage:
    minio:
      deploy: true
      image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
  mlpipelineUI:
    image: quay.io/opendatahub/ds-pipelines-frontend:latest
(base) [ddalvi@localhost data-science-pipelines-operator]$ oc get routes
NAME                    HOST/PORT                                                                                   PATH   SERVICES                PORT    TERMINATION          WILDCARD
ds-pipeline-sample      ds-pipeline-sample-dsp-envoy-enabled.apps.rosa.ddalvi-latest.al5q.p3.openshiftapps.com             ds-pipeline-sample      oauth   reencrypt/Redirect   None
ds-pipeline-ui-sample   ds-pipeline-ui-sample-dsp-envoy-enabled.apps.rosa.ddalvi-latest.al5q.p3.openshiftapps.com          ds-pipeline-ui-sample   8443    reencrypt/Redirect   None
(base) [ddalvi@localhost data-science-pipelines-operator]$

**Comment by dsp-developers** [comment]
PR was re-opened.
  A new image has been built to help with testing out this PR: `quay.io/opendatahub/data-science-pipelines-operator:pr-639`
An OCP cluster where you are logged in as cluster admin is required.

To use this image run the following:

```bash
cd $(mktemp -d)
git clone git@github.com:opendatahub-io/data-science-pipelines-operator.git
cd data-science-pipelines-operator/
git fetch origin pull/639/head
git checkout -b pullrequest 8011be4c6f199d513e7cbf82ae3edef87d53bb94
oc new-project opendatahub
make deploy IMG="quay.io/opendatahub/data-science-pipelines-operator:pr-639"
```

More instructions [here](https://github.com/opendatahub-io/data-science-pipelines-operator#deploy-dsp-instance) on how to deploy and test a Data Science Pipelines Application.

**Comment by hbelmiro** [comment]
> @hbelmiro tried deploying these changes for a couple of scenarios, but I see this warning when deploying:
> 
> ```
> oc apply -f dsp-envoy-disabled.yaml
> Warning: unknown field "spec.mlmd.envoy.deployRoute"
> datasciencepipelinesapplication.datasciencepipelinesapplications.opendatahub.io/sample created
> ```
> 
> For case 1 for instance, I just deployed case1, v2 with envoy enabled, but I don't see mlmd route coming up:
> 
> ```
> cat dsp-envoy-enabled.yaml 
> apiVersion: datasciencepipelinesapplications.opendatahub.io/v1alpha1
> kind: DataSciencePipelinesApplication
> metadata:
>   name: sample
> spec:
>   dspVersion: v2
>   mlmd:
>     deploy: true
>     envoy:
>       deployRoute: true
>   objectStorage:
>     minio:
>       deploy: true
>       image: 'quay.io/opendatahub/minio:RELEASE.2019-08-14T20-37-41Z-license-compliance'
>   mlpipelineUI:
>     image: quay.io/opendatahub/ds-pipelines-frontend:latest
> (base) [ddalvi@localhost data-science-pipelines-operator]$ oc get routes
> NAME                    HOST/PORT                                                                                   PATH   SERVICES                PORT    TERMINATION          WILDCARD
> ds-pipeline-sample      ds-pipeline-sample-dsp-envoy-enabled.apps.rosa.ddalvi-latest.al5q.p3.openshiftapps.com             ds-pipeline-sample      oauth   reencrypt/Redirect   None
> ds-pipeline-ui-sample   ds-pipeline-ui-sample-dsp-envoy-enabled.apps.rosa.ddalvi-latest.al5q.p3.openshiftapps.com          ds-pipeline-ui-sample   8443    reencrypt/Redirect   None
> (base) [ddalvi@localhost data-science-pipelines-operator]$ 
> ```
> 
> Looks like the field you added isn't being recognized, could you please deploy and re verify if you're seeing the same?

@DharmitD I retested scenario 1 and it works as expected.

```bash
oc get routes          
NAME                    HOST/PORT                                                                PATH   SERVICES                            PORT           TERMINATION          WILDCARD
ds-pipeline-md-sample   ds-pipeline-md-sample-kubeflow.apps.hbelmiro-10.dev.datahub.redhat.com          ds-pipeline-metadata-envoy-sample   oauth2-proxy   reencrypt/Redirect   None
ds-pipeline-sample      ds-pipeline-sample-kubeflow.apps.hbelmiro-10.dev.datahub.redhat.com             ds-pipeline-sample                  oauth          reencrypt/Redirect   None
ds-pipeline-ui-sample   ds-pipeline-ui-sample-kubeflow.apps.hbelmiro-10.dev.datahub.redhat.com          ds-pipeline-ui-sample               8443           reencrypt/Redirect   None
```

Due to the warning you see and the route not being created, I would bet you don't have the right version of DSPO deployed. Do you mind removing the namespaces and redeploying DSPO and the DSPA again?

**Comment by diegolovison** [comment]
Hello, if disabling the route is going to break the UI. Why are we spending trying to disable it?

**Comment by hbelmiro** [comment]
> Hello, if disabling the route is going to break the UI. Why are we trying to disable it?

@diegolovison it is meant just for dev testing.

**Pull request merged**